### ISSUE TITLE
Lookahead bias?

### ISSUE THREAD

**Issue by personal-coding** [kah-ve/MarketGAN#5]
You're using a negative shift in your data, doesn't this incorporate future data into your training set? `shift(-num_historical_days)`

**Comment by jjphung** [comment]
I don't think there is look ahead bias. At that stage, they are merely setting up training data for the model to try and predict `num_historical_days` ahead, which is the whole point of this repo. Furthermore, the next lines "Hold out the last year of trading for testing", specifically avoids look ahead bias by testing only future data.

**Comment by personal-coding** [comment]
Holding out data for testing data doesn't prevent look ahead bias when the data being held out has future data in it.

For instance, future data is being shifted backwards into the dataframe before the data is being split. The `.shift(-num_historical_days)` is taking future data and shifting it backwards.

Separately, when I see 90%+ accuracy related to stock data, there is likely look ahead bias somewhere.

```
for file in files:
            print(file)
            #Read in file -- note that parse_dates will be need later
            df = pd.read_csv(file, index_col='timestamp', parse_dates=True)
            df = df[['open','high','low','close','volume']]
            # #Create new index with missing days
            # idx = pd.date_range(df.index[-1], df.index[0])
            # #Reindex and fill the missing day with the value from the day before
            # df = df.reindex(idx, method='bfill').sort_index(ascending=False)
            #Normilize using a of size num_historical_days
            df = ((df -
            df.rolling(num_historical_days).mean().shift(-num_historical_days))
            /(df.rolling(num_historical_days).max().shift(-num_historical_days)
            -df.rolling(num_historical_days).min().shift(-num_historical_days)))
            #Drop the last 10 day that we don't have data for
            df = df.dropna()
            #Hold out the last year of trading for testing
            #Padding to keep labels from bleeding
            df = df[400:]
            #This may not create good samples if num_historical_days is a
            #mutliple of 7
            for i in range(num_historical_days, len(df), num_historical_days):
                self.data.append(df.values[i-num_historical_days:i])
```

**Comment by kah-ve** [comment]
Firstly, I suggest looking at the sources I present in the readme for better discussion about the training of the GANs. Namely this [repo](https://github.com/nmharmon8/StockMarketGAN). This was me trying to get accustomed to GANs and the value I added was just setting it up in Google Colab in an easy to use pipeline and some other modest changes if I remember correctly. 

It seems you are right. If the market data is listed from oldest to latest, then this would be forward bias since you are normalizing according to future values. If the market data is listed from latest to oldest, then in that case it would be just normalizing according to the past num of days as expected.

Now I see that the csv I shared has them in the oldest to latest format. Did I sort them later, I don't remember. Maybe, maybe not. Either way this was just a step into getting familiar with GANs for another project :)

**Comment by kah-ve** [comment]
I have updated the code so that it is sorted correctly from newest to oldest and this issue should be resolved. Also this model is not close to 90% accurate, that's just prediction of downs. But predicting downs accurately while being very inaccurate with predicting ups gives less confidence about the model.
